### ISSUE TITLE
Index/dimension order not preserved when going from and to DataFrame

### ISSUE THREAD

**Issue by samwisehawkins** [pydata/xarray#780]
If I convert  DataFrame --> Dataset -->  DataFrame, then the index ordering gets switched.  Is this a bug or a feature?

```
ids = ['A','B']
heights = [1,2]
times = pd.date_range('2010-01-01', '2010-01-02', freq='1D')
names = ['id','height','time']
index = pd.MultiIndex.from_product([ids, heights, times], names=names)
f1 = pd.DataFrame(index=index)            # has ordering id, height time
f1['V1'] = np.random.ranf(size=len(f1))
f1['V2'] = np.random.ranf(size=len(f1))

ds = xr.Dataset.from_dataframe(f1)  # has ordering id, height, time
f2 = ds.to_dataframe()             # has ordering height, id, time
```

**Comment by samwisehawkins** [comment]
This can easily be corrected using `reorder_levels`, but would be nice if this was automatic.

**Comment by samwisehawkins** [comment]
Moreover, I've just realised that the resulting DataFrame has a different ordering of the rows.  So as far as I can tell, in order to get back the same DataFrame in a round trip to and from a Dataset, you need to to reorder the levels of the column index, then resort the rows according to that column index. 

I'm doing this round trip as I'm storing DataFrames as netcdf files, and want to check that I can get do the round trip properly.

**Comment by shoyer** [comment]
So I'm actually not sure whether to call this a bug or a feature. But I can explain why it works this way and maybe we can come up with something better.

With `DataArray.to_series()`, we are indeed careful to output the hierarchical index in the same order as the array dimensions. So it works there.

But on a Dataset, we don't necessarily have a unique ordering for the dimensions, because in general (though somewhat rarely in practice) the ordering of dimensions can differ between variables. This is why `Dataset.dims` returns a `SortedKeysDict` -- to avoid any implicit state derived off the order in which dimensions were added.

When converting a DataFrame, we currently build the MultiIndex independently of the data variables, so somewhat logically we simply take dimensions in sorted order. It might make more sense, though, to instead order levels in order of appearance on Dataset (non-index?) variables. I do try to avoid making heuristic choices like this, though, which is why it didn't make it into xarray already.

This code is pretty self-contained if you want to experiment and/or put together a PR:
https://github.com/pydata/xarray/blob/v0.7.1/xarray/core/dataset.py#L1858-L1872

Basically, you need to ensure that `ordered_dims` is an `OrderedDict` with keys in the order you want for the resulting DataFrame.

**Comment by samwisehawkins** [comment]
Thanks, will have a look.
